biz.aQute.repository/src/aQute/bnd/repository/maven/pom/provider/BndPomRepository.java

```diff
@@ -75,21 +80,90 @@
 	private String				queryUrl;
 	private ScheduledFuture<?>	pomPoller;
 
-	public synchronized void init() {
+	private String				status;
+
+	private PromiseFactory		promiseFactory		= new PromiseFactory(Executors.newSingleThreadExecutor());
+
+	@SuppressWarnings("deprecation")
+	private boolean init() {
 		try {
-			if (inited)
-				return;
-			inited = true;
+			if (initialized == null) {
+				prepare();
+			}
+			return initialized.getValue();
+		} catch (Exception e) {
+			Exceptions.duck(e);
+		}
+		return false;
+	}
+
+	@SuppressWarnings("deprecation")
+	@Override
+	public void prepare() throws Exception {
+		if (configuration.name() == null) {
+			status("Must get a name");
+		}
+
+		this.name = configuration.name();
 
+		if (configuration.snapshotUrl() != null && configuration.snapshotUrls() != null) {
+			status("snapshotUrl and snapshotUrls property is set. Please only use snapshotUrl.");
+		}
+
+		if (configuration.releaseUrl() != null && configuration.releaseUrls() != null) {
+			status("releaseUrl and releaseUrls property is set. Please only use releaseUrl.");
+		}
+
+		if (configuration.pom() != null) {
+			pomFiles = Strings.split(configuration.pom())
+				.stream()
+				.map(part -> {
+					File f = IO.getFile(part);
+					return f.isFile() ? f.toURI() : URI.create(part);
+				})
+				.collect(toList());
+			if (pomFiles.isEmpty()) {
+				status("Pom is neither a file nor a revision " + configuration.pom());
+			}
+		} else if (configuration.revision() != null) {
+			revisions = Strings.split(configuration.revision())
+				.stream()
+				.map(Revision::valueOf)
+				.filter(Objects::nonNull)
+				.collect(toList());
+			if (revisions.isEmpty()) {
+				status("Revision is neither a file nor a revision " + configuration.revision());
+			}
+		} else if (configuration.query() != null) {
+			this.query = configuration.query();
+			this.queryUrl = configuration.queryUrl("http://search.maven.org/solrsearch/select");
+		} else {
+			status("Neither pom, revision nor query property are set");
+		}
+
+		initialized = promiseFactory.submit(this::internalInitialize);
+	}
+
+	@SuppressWarnings("deprecation")
+	private boolean internalInitialize() {
+		if (!isOk()) {
+			return false;
+		}
+		try {
 			Workspace workspace = registry.getPlugin(Workspace.class);
 			HttpClient client = registry.getPlugin(HttpClient.class);
 			localRepo = IO.getFile(configuration.local(MAVEN_REPO_LOCAL));
 			File location = workspace.getFile(getLocation());
 
-			List<MavenBackingRepository> release = MavenBackingRepository.create(configuration.releaseUrls(), reporter,
-				localRepo, client);
-			List<MavenBackingRepository> snapshot = MavenBackingRepository.create(configuration.snapshotUrls(),
-				reporter, localRepo, client);
+			String releaseUrl = configuration.releaseUrl() != null ? configuration.releaseUrl()
```

Comment:
you should do all this work in prepare() since it is only done once.

Reply:
I had it in the prepare method in a `() -> {}` block, but the method became to long, so I thought it might be better to split it. prepare is doing the checks and kicks of the init Promise and the extra method is doing the extra initialization. I can do it in one method, but it looks ugly though.